Add EIP-5875: Instruction to Get Transaction Index of Block (#5875)
ethereum/EIPs: EIPS/eip-5875.md

## Specification

The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "NOT RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as described in RFC 2119 and RFC 8174.

1. Parameter Table

* `FORK_NUM_TBD`: the fork block number to start introducing this change.
* `OPCODE_TBD`: the opcode value being introduced.

2. On `TBD_FORK_NUM`, a new instruction `TXINDEX (0xXX)` is introduced

3. Upon execution, the instruction pushes a value of the index of the current transaction in its block in the stack.

4. The gas cost of `TXINDEX` is specified as `2`

## Rationale

1. The bound of size: It's unlikely the transaction index could ever grow close to the 2^256-1. When restricting bound of the transaction index it could potentially help some computations, for example of ChainId being used in [EIP-712](./eip-712.md) imposing a computational ceiling. But for simplicity we choose to not limit the size of the transaction index so itdefaults_ to a maximum of `2^256-1`.

2. The gas cost of `TXINDEX` is picked to be the same as `NUMBER` and `TIMESTAMP`.

### Further discussions needed

> 4. Further discussion is needed about potentially bounding the transaction index for simpler on-chain computation if use case arises.
> 5. Further discussion is needed about whether to use transaction index within the block or globally. Globally makes it easier to compute number distance between two transactions directly but is useless for DAG
> if we foresee a DAG future.

Reviewer: ```suggestion
- Further discussion is needed about potentially bounding the transaction index for simpler on-chain computation if use case arises.
- Further discussion is needed about whether to use transaction index within the block or globally. Globally makes it easier to compute number distance between two transactions directly but is useless for DAG
if we foresee a DAG future.
```

---
eip: 5875

> title: Opcode for TX Number

Reviewer 2: Generally "transaction index" is the accepted way of referring to a certain transaction within a block / the numerical identifier of the transaction in the block.

> description: Add an opcode to retrieve TX number

Reviewer 2: ```suggestion
description: Add an opcode to retrieve transaction index in block
```

> description: Add an opcode to retrieve transaction index in block

Reviewer: ```suggestion
description: Add an opcode to retrieve the transaction index in a block
```

author: Zainan Victor Zhou (@xinbenlv)
discussions-to: https://ethereum-magicians.org/t/eip-5875-opcode-for-transaction-number-in-a-block/11612
status: Draft
type: Standards Track
category: Core
created: 2022-11-04
---

## Abstract

> This EIP adds an instruction `TXINDEX` to retrieve the Transaction Index inside of a block.

Reviewer: ```suggestion
This EIP adds an instruction, `TXINDEX`, to retrieve the index of the current transaction in a block.
```

> Upon `FORK_NUM_TBD`, add EVM an opcode instruction with value `OPCODE_TBD` and name `TXNUM` to retrieve TX number inside of a block.

Reviewer: ```suggestion
This EIP adds an opcode instruction, `OPCODE_TBD (TXNUM)` to retrieve the transaction number inside of a block.
```

> This EIP adds an instruction, `OPCODE_TBD (transactionNUM)` to retrieve the Transaction Index inside of a block.

Reviewer 2: ```suggestion
This EIP adds an instruction `TXINDEX` to retrieve the Transaction Index inside of a block.
```

Not sure the need for the TBD mnemonic here. Please also choose a specific opcode number. It can be changed if this EIP is considered for an upgarde.

## Motivation

> A TX is an atomic point of time for world state.

Reviewer: ```suggestion
A transaction is an atomic point of time for a world state.
```

> Currently we have access to blocknum but there is no way to access a tx number within a smart contract.

Reviewer: ```suggestion
Currently, we have access to the block number through `BLOCKNUMBER (0x43)` but there is no way to access a transaction number within a smart contract.
```

> We hereby propose a TXNUM opcode instruction to fetch TXNUM which can be used to

Reviewer: ```suggestion
This EIP proposes an opcode to fetch the transaction number, which can be used to:
```

> A transaction is an atomic point of time for a world state.
> 
> Currently, we have access to the block number through `BLOCKNUMBER (0x43)` but there is no way to access a Transaction Index within a smart contract.
> 
> This EIP proposes an opcode to fetch the Transaction Index, which can be used to:

Reviewer: ```suggestion
A transaction is an atomic point of time for a world state. Currently, we have access to the block number through `BLOCKNUMBER (0x43)`, but there is no way to access the transaction index within a smart contract. This EIP adds an opcode to fetch the transaction index, which can be used to:
```

> identify the ordering of exact transaction in its block
> Better than blockNum: identifying uniquely and unambiguously the world state.
> Better than txHash: allow simply arithmatic by BlockNum.TxNum
> It would be helpful for on-chain and off-chain indexing and snapshotting.

Reviewer: ```suggestion
* Identify the exact ordering of transactions within a block
* Uniquely and unambiguously identify the world state (better than `BLOCKNUMBER`)
* Allow simple arithmetic on the transaction number (better than an opcode to get the transaction hash)
* On-chain and off-chain indexing and snapshotting
```

> An example of use-case is in snapshot weighted voting, we want to store snapshot of voting rights. The current option with with either `block.timesamp` or `block.number` all only have the granularity at the level of blocks. But world state changes across transaction. If there is any TX in the same block cause a voting right to change, the blocknum or `block.timstamp` will not be sufficient to pin-point the world state of snapshot of voting rights.

Reviewer: ```suggestion
An example use-case is snapshot weighted voting, in which a snapshot of voting rights is generated. The current options are with `block.timesamp` or `block.number`, which only have block-level granularity. But the world state changes across transactions within a block. If there is any transaction in the same block that causes a voting right to change, the block number or block timestamp is not sufficient to pinpoint the world state of a snapshot of voting rights.
```

> ## Specification

Reviewer 2: You don't specify how much the operation costs.

Author: Good point. Added now. Thank you!

The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "NOT RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as described in RFC 2119 and RFC 8174.

1. Parameter Table

* `FORK_NUM_TBD`: the fork block number to start introducing this change.

> * `OPCODE_TBD`: the opcode value being introduced.

Reviewer 2: This is unnecessary.

Author: Sure, I could remove it.

QQ: If we don't make OPCODE a placeholder number, how do we allow people to propose EIPs picking arbitrary unused new opcodes without conflicting with other drafts, knowing some of such core EIPs are taking long / never going to be adopted?

> 1. Upon `TBD_FORK_NUM`, add EVM an opcode with name `TXNUM` and value `TBD_OPCODE` instruction to retrieve TX number.

Reviewer 2: ```suggestion
1. Upon `TBD_FORK_NUM`, add EVM an opcode with name `TXNUM` and value `TBD_OPCODE` instruction to retrieve TX number.
```
This sentence is confusing.

Author: I updated the wording, can you see if it improves?

> 2. Upon `TBD_FORK_NUM`, A new instruction is added to EVM with name `TX_NUM` and value `TBD_OPCODE` instruction to retrieve transaction index.

Reviewer 2: ```suggestion
2. On `TBD_FORK_NUM`, a new instruction `TXINDEX (0xXX)` is introduced

> 2. The returned value is in the length of `uint256`, a word length of EVM.

Reviewer 2: Returned is the wrong verb. The result is placed on the stack. It's also unnecessary to specify the size because all stack elements are 256-bit.

Author: Done

> 3. Upon execution the instruction pushes a value of the index of current transaction in its block in the stack.

Reviewer: ```suggestion
3. Upon execution, the instruction pushes a value of the index of the current transaction in its block in the stack.
```

## Rationale

> 1. We choose the Transaction Index length for a world length for simplicity.

Reviewer 2: I don't understand this sentence. Did you mean word length? In that case I would just omit this because the word length is not decidable by an instruction.

Author: I was referring to the bounding. Updated the sentence and could you check if it reads better now?

> 2. While currently most block contains O(100) transactions, we didn't impose a boundary for Transaction Index other than uint256 to allow

Reviewer 2: You don't need Big O notation here, just state that typical number of transactions in the block.

```suggestion
2. While currently most block contains O(100) transactions, we didn't impose a boundary for transaction index other than uint256 to allow
```

a future where each block can have much larger blocks.

### Further discussions needed

> 4. further discussion is needed about potentially bounding the TX num for simpler onchain computation if use case arises.

Reviewer 2: Capitalize these please.

> 4. Further discussion is needed about potentially bounding the transaction num for simpler onchain computation if use case arises.

Reviewer 2: ```suggestion
4. Further discussion is needed about potentially bounding the transaction index for simpler on-chain computation if use case arises.
```

> 5. Further discussion is needed about whether to use Transaction Index within the block or globally. Globally makes it easier to compute number distance between two transactions directly but is useless for DAG

Reviewer 2: ```suggestion
5. Further discussion is needed about whether to use transaction index within the block or globally. Globally makes it easier to compute number distance between two transactions directly but is useless for DAG
```
